nixos/modules/services/misc/dwm-status.nix

```diff
@@ -0,0 +1,70 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.dwm-status;
+
+  order = concatMapStringsSep "," (feature: ''"${feature}"'') cfg.order;
+
+  configFile = pkgs.writeText "dwm-status.toml" ''
+    order = [${order}]
+
+    ${cfg.extraConfig}
+  '';
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    services.dwm-status = {
+
+      enable = mkEnableOption "dwm-status user service";
+
+      order = mkOption {
+        type = types.listOf (types.enum [ "audio" "backlight" "battery" "cpu_load" "network" "time" ]);
+        description = ''
+          List of enabled features in order.
+        '';
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Extra config in TOML format.
+        '';
+      };
+
+    };
+
+  };
+
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+
+    services.upower.enable = elem "battery" cfg.order;
+
+    systemd.user.services.dwm-status = {
+      description = "Highly performant and configurable DWM status service";
+      wantedBy = [ "graphical-session.target" ];
+      partOf = [ "graphical-session.target" ];
+
+      serviceConfig.ExecStart =
+        let
+          package = pkgs.dwm-status.override {
+            enableAlsaUtils = elem "audio" cfg.order;
+            enableNetwork = elem "network" cfg.order;
```

Comment:
Actually, I don't think this is needed, because these two settings are `true` by default. Currently if the user doesn't include them in `order`, dwm-status would be recompiled just to get a version without support for those, even though there would be a version in the cache with support.

So unless this has some deeper meaning behind it, just using `"${pkgs.dwm-status}/bin/dwm-status ..."` should be fine

Reply:
Closure sizes for enabled/disabled flags are

| flags enabled | Size |
| --- | --- |
| false | 65.2MB | 
| true | 99.9MB |

So having the flags turned on adds about 34.7MB / 53.2% to the closure size, which should be alright, given that people won't have to compile it themselves.

Reply:
The reason for these switches is that only the necessary dependencies accoring to the config are installed. I wanted to keep the closure size to a minimum. Perhaps I could add the package in every possible configuration in `all-packages.nix` or add an option in this module if the minimal package should be used despite the fact of recompiling the package or just use the cached version

Reply:
On my system, building dwm-status fetches 447MiB of dependencies and builds on mostly 800% CPU for 2m26s, which IMO is much worse than quietly downloading 34.7MiB extra. I think it might be better if the default included everything, but offered a `package` option so that a dwm-status with a reduced feature-set can be provided. This auto-detection may not be obvious to users who haven't read the module.

Reply:
Hmm, I don't think adding an option for that is reasonable. And adding all 4 configurations to all-packages.nix also doesn't sound very nice.

A good solution (which needs more discussion though), would be to have a global NixOS option like `system.minimalClosure`, which all modules could read and optionally do more compilation for some smaller closure sizes.

For now I think it's probably alright to leave it like this, even though dwm-status is a rust package, it's not as bad as other ones with build times.

Reply:
@Infinisil Why is a `package` option unreasonable? There are around 165 NixOS options named package, mostly for various services

Reply:
Thank you guys for the input. The `package` option seems to be a nice fit for this.